modules/index/src/main/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskController.java

```diff
@@ -153,46 +160,84 @@ private void onIndexRemoved(RemoveIndexEventParameters parameters) {
-
             if (isLocalNode(clusterService, parameters.leaseholderId())) {
-                if (localNodeIsPrimaryReplicaForTableIds.add(tableId)) {
-                    scheduleTasksOnPrimaryReplicaElectedBusy(tableId);
-                }
+                scheduleTasksOnPrimaryReplicaElectedBusy(primaryReplicaId);
             } else {
-                if (localNodeIsPrimaryReplicaForTableIds.remove(tableId)) {
-                    changeIndexStatusTaskScheduler.stopTasksForTable(tableId);
-                }
+                scheduleStopTasksOnPrimaryReplicaElected(primaryReplicaId);
             }
         });
     }
 
-    private void scheduleTasksOnPrimaryReplicaElectedBusy(int tableId) {
+    private void scheduleTasksOnPrimaryReplicaElectedBusy(PartitionGroupId partitionGroupId) {
+        // It is safe to get the latest version of the catalog because the PRIMARY_REPLICA_ELECTED event is handled on the metastore thread.
+        Catalog catalog = catalogService.catalog(catalogService.latestCatalogVersion());
+
+        IntListIterator tableIds =
+                getTableIdsForPrimaryReplicaElected(catalog, partitionGroupId, localNodeIsPrimaryReplicaForTableIds::add);
+
+        while (tableIds.hasNext()) {
+            for (CatalogIndexDescriptor indexDescriptor : catalog.indexes(tableIds.nextInt())) {
+                switch (indexDescriptor.status()) {
+                    case REGISTERED:
+                        changeIndexStatusTaskScheduler.scheduleStartBuildingTask(indexDescriptor);
+
+                        break;
+
+                    case STOPPING:
+                        changeIndexStatusTaskScheduler.scheduleRemoveIndexTask(indexDescriptor);
+
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+        }
+    }
+
+    private void scheduleStopTasksOnPrimaryReplicaElected(PartitionGroupId partitionGroupId) {
         // It is safe to get the latest version of the catalog because the PRIMARY_REPLICA_ELECTED event is handled on the metastore thread.
         Catalog catalog = catalogService.catalog(catalogService.latestCatalogVersion());
 
-        for (CatalogIndexDescriptor indexDescriptor : catalog.indexes(tableId)) {
-            switch (indexDescriptor.status()) {
-                case REGISTERED:
-                    changeIndexStatusTaskScheduler.scheduleStartBuildingTask(indexDescriptor);
+        IntListIterator tableIds =
+                getTableIdsForPrimaryReplicaElected(catalog, partitionGroupId, localNodeIsPrimaryReplicaForTableIds::remove);
+
+        while (tableIds.hasNext()) {
+            changeIndexStatusTaskScheduler.stopTasksForTable(tableIds.nextInt());
+        }
+    }
 
-                    break;
+    private IntListIterator getTableIdsForPrimaryReplicaElected(
+            Catalog catalog,
+            PartitionGroupId partitionGroupId,
+            Int2BooleanFunction filter
```

Comment:
I don't really like the approach with a filter that changes the internal state. Let's make the filter simple, but then immediately delete/add table IDs in batch, it looks clearer and more obvious.

Reply:
Done